### PR TITLE
[Kernel] Support deep_gemm for linear methods

### PR DESCRIPTION
**Pull request by artetaout**
##  Insight
When using Qwen3-32B-FP8, the decode stage is using cutlass FP8, which makes it slower than sglang
* vLLM 27ms 
image
image

* sglang 14ms
image
image

And as I checked, vLLM use DeepGEMM in MOE, but not in FP8LinearMethod
So I add an option to use DeepGEMM and JIT as sglang does

## Accuracy Result
* origin cutlass FP8 decode

`VLLM_USE_DEEP_GEMM=0 lm_eval --model vllm --model_args pretrained=/mnt/Qwen3-32B-FP8 --tasks gsm8k --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6224|±  |0.0134|
|     |       |strict-match    |     5|exact_match|↑  |0.7445|±  |0.0120|

* our DeepGEMM FP8 decode

`VLLM_USE_DEEP_GEMM=1 lm_eval --model vllm --model_args pretrained=/mnt/Qwen3-32B-FP8 --tasks gsm8k --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6346|±  |0.0133|
|     |       |strict-match    |     5|exact_match|↑  |0.7430|±  |0.0120|

## Performance Result

* origin cutlass FP8 decode

`VLLM_USE_V1=1 VLLM_USE_DEEP_GEMM=0 vllm serve Qwen/Qwen3-32B-FP8/  --no-enable-prefix-caching`

```
============ Serving Benchmark Result ============
Successful requests:                     10
Benchmark duration (s):                  4.56
Total input tokens:                      1000
Total generated tokens:                  1000
Request throughput (req/s):              2.19
Output token throughput (tok/s):         219.24
Total Token throughput (tok/s):          438.48
---------------Time to First Token----------------
Mean TTFT (ms):                          291.90
Median TTFT (ms):                        318.24
P99 TTFT (ms):                           319.53
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          43.06
Median TPOT (ms):                        42.83
P99 TPOT (ms):                           44.89
---------------Inter-token Latency----------------
Mean ITL (ms):                           43.06
Median ITL (ms):                         42.89
P99 ITL (ms):                            43.39
==================================================
```
* our DeepGEMM FP8 decode

`VLLM_USE_V1=1 VLLM_USE_DEEP_GEMM=1 vllm serve Qwen/Qwen3-32B-FP8/  --no-enable-prefix-caching`

```
============ Serving Benchmark Result ============
Successful requests:                     10
Benchmark duration (s):                  2.51
Total input tokens:                      1000
Total generated tokens:                  1000
Request throughput (req/s):              3.98
Output token throughput (tok/s):         398.03
Total Token throughput (tok/s):          796.07
---------------Time to First Token----------------
Mean TTFT (ms):                          287.71
Median TTFT (ms):                        314.18
P99 TTFT (ms):                           315.46
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          22.43
Median TPOT (ms):                        22.18
P99 TPOT (ms):                           24.44
---------------Inter-token Latency----------------
Mean ITL (ms):                           22.43
Median ITL (ms):                         22.24
P99 ITL (ms):                            22.69
==================================================
```

* to reproduce
```
python3 benchmark_serving.py  --model /data/Qwen3-32B-FP8 --num-prompts 10 --dataset-name random --random-input-len 100 --random-output-len 100 --ignore-eos
```

* need to pre-install DeepGEMM
```
git clone https://github.com/deepseek-ai/DeepGEMM.git --recursive
cd DeepGEMM && git checkout bfe983c
python setup.py install
```

* support for vLLM V1 and torch.compile is done

### CODE REVIEW

**Comment by artetaout** [comment]
hi, I've fixed lint and please give some review advice ! @mgoin @robertgshaw2-redhat @tlrmchlsmth

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 1]
```diff
@@ -2,12 +2,15 @@
...
 import functools
+import importlib.util
 import json
 import os
 from typing import Any, Callable, Optional, Union
 
 import torch
 
+import vllm.envs as envs
+import vllm.model_executor.layers.quantization.deepgemm as deepgemm
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py]: Could you move this import into the conditional in `apply_w8a8_block_fp8_linear`?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 2]
```diff
@@ -114,6 +118,33 @@ def apply_w8a8_block_fp8_linear(
     input_2d = input.view(-1, input.shape[-1])
     output_shape = [*input.shape[:-1], weight.shape[0]]
 
+    output_dtype = input.dtype
+    dtype_supported = output_dtype == torch.bfloat16
+    shape_supported = weight.shape[0] % 128 == 0 and weight.shape[1] % 128 == 0
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py]: Make these names more specific e.g. dg_supported_dtype.
Also doesn't deepgemm only support Hopper GPUs? So we should be checking compute capability

**Comment by mgoin** [comment]
> currently ONLY support vLLM V0

Why is this the case and could you support V1? We will deprecate V0 soon

For your benchmarking, what GPU did you use?

**Comment by artetaout** [comment]
> > currently ONLY support vLLM V0
> 
> Why is this the case and could you support V1? We will deprecate V0 soon
> 
> For your benchmarking, what GPU did you use?

1. Currently, It does not support torch.compile. But we will support V1 and torch.compile later
2. use H20

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 1 above)
Comment (artetaout) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py]: OK, I will fix

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 2 above)
Comment (artetaout) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py]: OK, I will fix

File: vllm/model_executor/layers/quantization/deepgemm.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    assert triton.cdiv(N, block_n) == Bs.shape[0]
+    assert triton.cdiv(K, block_k) == Bs.shape[1]
+
+    C_shape = A.shape[:-1] + (N, )
+    C = A.new_empty(C_shape, dtype=output_dtype)
+
+    return M, N, K, C
+
+
+def w8a8_block_fp8_matmul_deepgemm(
```
Comment (youkaichao) [inline vllm/model_executor/layers/quantization/deepgemm.py:49]: to support `torch.compile`, i think you can wrap the function inside a custom op, examples here:

https://github.com/vllm-project/vllm/blob/5f2cd251d212eed3052c5406875eb26811335d3e/vllm/attention/layer.py#L418

Comment (artetaout) [reply vllm/model_executor/layers/quantization/deepgemm.py:49]: thx! we will support it soon!

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 1 above)
Comment (artetaout) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py]: done and commit is squashed

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 2 above)
Comment (artetaout) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py]: done and squashed

**Comment by artetaout** [comment]
> > currently ONLY support vLLM V0
> 
> Why is this the case and could you support V1? We will deprecate V0 soon
> 
> For your benchmarking, what GPU did you use?

Support V1 is done, and result for performance and accuracy is updated ! 

Besides, some comments on code are also fixed !

File: vllm/model_executor/layers/quantization/deepgemm.py (hunk 3 above)
Comment (artetaout) [reply vllm/model_executor/layers/quantization/deepgemm.py:49]: The torch.compile is SUPPORTED. the result is updated as well

File: vllm/model_executor/layers/quantization/deepgemm.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+    kernel_type = DeepGemmKernelType.GROUPED_GEMM_NT_F8F8BF16_CONTIG
+    _maybe_compile_deep_gemm_one_type_all(kernel_type, n, k, num_groups)
+
+    with _log_jit_build(m, n, k, kernel_type):
+        deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_contiguous(
+            lhs, rhs, out, m_indices)
+
+
+def gemm_nt_f8f8bf16(
```
Comment (bnellnm) [inline vllm/model_executor/layers/quantization/deepgemm.py]: Can you explain how this method is different than `deep_gemm.gemm_fp8_fp8_bf16_nt`?  There's a lot of lower level code copied from DeepGemm but it's not immediately obvious why it is necessary.

Comment (artetaout) [reply vllm/model_executor/layers/quantization/deepgemm.py]: In sglang, It provides command for pre-compile DeepGEMM, to save time during start stage.
And since now this commit does not plan to  support pre-compile commands, so I've removed them.
Support for vllm pre-compile will be in another commit .

Comment (artetaout) [reply vllm/model_executor/layers/quantization/deepgemm.py]: > Can you explain how this method is different than `deep_gemm.gemm_fp8_fp8_bf16_nt`? There's a lot of lower level code copied from DeepGemm but it's not immediately obvious why it is necessary.

In sglang, It provides command for pre-compile DeepGEMM, to save time during start stage.
And since now this commit does not plan to support pre-compile commands, so I've removed them.
Support for vllm pre-compile will be in another commit .

**Comment by shixianc** [comment]
Does this deepgemm kernel only work for block_quant (fine-grained) fp8 models? Can it work with general per-tensor/channel quant fp8 models?

**Comment by artetaout** [comment]
> Does this deepgemm kernel only work for block_quant (fine-grained) fp8 models? Can it work with general per-tensor/channel quant fp8 models?

I've never try a channel/tensor-wise FP8 model, maybe you can provide one or two.

**Comment by shixianc** [comment]
@artetaout 
I'm testing on our own model that quantized with AutoFP8 (per-tensor fp8)
Here's one on HF: RedHatAI/Mixtral-8x7B-Instruct-v0.1-AutoFP8

However I were looking at the deepgemm code https://github.com/deepseek-ai/DeepGEMM/blob/main/deep_gemm/jit_kernels/gemm.py#L167-L170 it seems the weight scale tensors must be in per-block/group. Does it mean it won't work if the input weight scales are per-tensor? 
I meant we could try expanding our per-tensor weight scales to the required shape, but do you think it still provide  speed improvement by DeepGEMM itself?

**Comment by artetaout** [comment]
> @artetaout I'm testing on our own model that quantized with AutoFP8 (per-tensor fp8) Here's one on HF: RedHatAI/Mixtral-8x7B-Instruct-v0.1-AutoFP8
> 
> However I were looking at the deepgemm code https://github.com/deepseek-ai/DeepGEMM/blob/main/deep_gemm/jit_kernels/gemm.py#L167-L170 it seems the weight scale tensors must be in per-block/group. Does it mean it won't work if the input weight scales are per-tensor? I meant we could try expanding our per-tensor weight scales to the required shape, but do you think it still provide speed improvement by DeepGEMM itself?

i think it will.

**Comment by artetaout** [comment]
hi reviewers!  thx for review! but the CI seems to be failed, but the failure has nothing to do with current deepgemm
 logic....   i wonder if i've missed something ?  or please tell me what to do next @mgoin @youkaichao @bnellnm

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 5]
```diff
@@ -97,6 +100,19 @@ def dispatch_w8a8_blockscale_func(
...
 
+def should_use_deepgemm(output_dtype: torch.dtype, weight: torch.Tensor):
+    """
+    Check if DeepGEMM should be used based on the output dtype and weight shape.
+    DeepGEMM is only supported for bfloat16 output dtype and weights with shape
+    divisible by 128.
+    """
+
+    return (current_platform.is_cuda()
+            and current_platform.has_device_capability(90) and has_deep_gemm
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py]: This should use the new `is_device_capability` function to check for exact match (you might need to merge from main), since Hopper is the only supported architecture for deepgeem

**Review by mgoin — COMMENTED** [review]
Nice simplification work, this looks much better! Just one nit and I'll help you look at the CI failures. We have some known failures

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 5 above)
Comment (artetaout) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py]: ok, will fix soon

**Comment by artetaout** [comment]
> Nice simplification work, this looks much better! Just one nit and I'll help you look at the CI failures. We have some known failures

fixed, and now it seems to be pending...

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 6]
```diff
@@ -114,6 +130,31 @@ def apply_w8a8_block_fp8_linear(
...
+        input_2d = input.view(-1, input.shape[-1])
+        output_shape = [*input.shape[:-1], weight.shape[0]]
+
+        q_input, x_scale = per_token_group_quant_fp8(
+            input_2d,
+            block_size[1],
+            column_major_scales=True,
+        )
+
+        output = deepgemm.w8a8_block_fp8_matmul_deepgemm(
```
Comment (shixianc) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py]: After registering custom op in vllm, I thought we need to call using `torch.ops.vllm.w8a8_block_fp8_matmul_deepgemm` instead of calling directly?

Comment (artetaout) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py]: YES!

**Comment by artetaout** [comment]
hi reviewers , is there anything i need to do more? @mgoin @shixianc

**Review by shixianc — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
LGTM, thank you for the contribution! We should aim to document this and possibly make a script/tool for installing deepgeem to use this in the future, similar to nixl (https://github.com/vllm-project/vllm/blob/da9b523ce1fd5c27bfd18921ba0388bf2e8e4618/tools/install_nixl.sh)

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by artetaout** [comment]
> LGTM, thank you for the contribution! We should aim to document this and possibly make a script/tool for installing deepgeem to use this in the future, similar to nixl (https://github.com/vllm-project/vllm/blob/da9b523ce1fd5c27bfd18921ba0388bf2e8e4618/tools/install_nixl.sh)

the document about DeepGEMM installation will be release soon...  by my colleague @HanHan009527
